<a href="https://colab.research.google.com/github/Tiger-A/Yandex_Data-Science/blob/master/pj_5_enter_in_ml.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Результат ревью:

Я отмечаю адекватный уровень твоей программистской подготовки. Ты используешь продвинутые концепции языка Python, библиотеки pandas и одной из основных библиотек машинного обучения sckit-learn. Это показывает твой уровень владения инструментами анализа данных и machine learning.

С комментированием никаких проблем нет. Твоих объяснений достаточно для понимания твоими коллегами хода мыслей. Ссылки из брифа проекта в нужных местах также помогают следовать за развитием событий и дают полную информацию касаемо того, что будет происходить в том или ином разделе проекта.

Мне было просто проверять твой проект, потому что он чётко структурирован, с использованием оглавления. 

Зелёным цветом я отметил и прокомментировал удачные и элегантные решения, на которые ты можешь опираться в будущих проектах.

Жёлтым цветом я выделил то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас.

Красных комментариев, которые бы не позволяли мне принять твой проект в его нынешнем состоянии - нет. 

ПРОЕКТ ПРИНЯТ!

Подводя итог: отличная работа, молодец! Желаю тебе успехов в учёбе, желаю, чтобы проекты сдавались или сразу, или после первой итерации! Удачи!

p.s. я тебе на твой адрес эл/почты сейчас вышлю некоторые материалы из своей электронной библиотеки - надеюсь они будут тебе полезными в учёбе.</div>

Приветствую коллега! 

Меня зовут Александр Непомнящий и я буду проверять твой проект. Предлагаю общаться на «ты» :))) Но, если это не удобно - мы вполне можем перейти на вы".

Вкратце обо мне: я работаю в сфере информационных технологий, аналитика (анализ) данных - это и мой рабочий функционал и личная, интересная мне область для саморазвития. Используемый стек: python, pandas, графические библиотеки matplotlib/seaborn/plotly. Сфера применения: маркетинговая аналитика в медицинской клинике (Клиника «Кивач», www.kivach.ru, республика Карелия). Если вдруг есть вопрос – что делает аналитик в коммерческой медицине – то вот короткий ответ. Самый главный вопрос, на который мы хотим получить ответ (и который в дальнейшем приведёт к перераспределению рекламного бюджета, плюс напрямую повлияет на разработку и предложению рынку новых лечебных программ) - это КТО ОН, НАШ КЛИЕНТ? Фактически это задача (на нашем внутреннем языке формулируется как построение и анализ «ПОРТРЕТА КЛИЕНТА»). Визуализируем и пытаемся придать смысл половозрастным данным наших клиентов, кто из какого региона и в какие времена года к нам приезжает. Учитываем количество приездов. Учитываем на какую программу этот конкретный клиент приехал в первый раз и как в дальнейшем меняются его предпочтения. Пытаемся построить модель оттока и спрогнозировать, когда клиент уйдёт в отток. Смотрим цифры (суммы трат клиентов на отдельных группах медицинских программ). Анализируем медицинские услуги, которые входят в путёвку, а также что, в каких количествах и на какую сумму клиент покупает дополнительно. Очень вкратце как-то так.

Теперь к сути: основная цель моей проверки - пройти по шагам, по логике твоего решения и обсудить те моменты, которые можно изменить, улучшить, довести до идеала. Такой формат работы двух программистов (аналитиков) ты или уже встречал(а), или встретишь на своей новой работе. Кто-то такой процесс называет "код-ревью", ну а как по мне, так это обмен опытом двух коллег, когда каждый из нас становиться лучше в процессе дискуссии.

Свои мысли, советы, комментарии и вопросы я буду оставлять в текстовых ячейках (markdown) ПОД твоим кодом. 
Для твоего удобства все эти заметки будут "цветными", как на примерах ниже:

"Зелеными" комментариями я буду отмечать твой элегантный ("красивый") код и/или удачные решения, на которые ты можешь смело опираться в будущих проектах.
<div class="alert alert-success">
<b>Отлично! 👍</b>

Отличная находка, элегантный код ...</div>

"Жёлтые" комментарии - это мои размышления, предложения, что-то, что на следующей итерации можно сделать по-другому. Твоя реакция на "жёлтые" комментарий необходима и подразумевает или правки, или реакцию или аргументированное согласие/не согласие. Я искренне верю, что наш двусторонний диалог под "желтым" комментарием - это то, от чего твой проект только выиграет. <div class="alert alert-warning">
<b>Комментарий 👉</b>

Здесь я бы предложил использовать .... Мои аргументы в пользу этого ...</div>

"Красные" комментарий будут появляться, когда что-то пойдёт не так: например, код будет выдавать ошибку. Или, когда в проекте не будет ответов на основные вопросы, ради которых весь анализ и выполнялся. "Красные" комментарии необходимо будет исправить, прежде чем я смогу принять проект.
<div class="alert alert-danger">
<b>Необходимо исправить ❌</b>

В этой ячейке код работает не так как задумано ...</div>

Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Это поможет мне оперативно выполнить повторную проверку твоего проекта.

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом, лучше всего в новой ячейке, сразу под моим комментарием. Если для своих комментариев ты будешь использовать какой-то заметный цвет, то мне будет легче отследить твои изменения, например так:

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Текст твоего комментария ...</font>
</div>

Ну и наконец информируй меня в том случае, если по твоему проекту подходит дедлайн! В таком случае я приложу усилия, чтобы ты сдал(а) его вовремя; буду оперативно, в приоритете проверять итерации ревью, а также отвечать на твои вопросы, если этого будет требовать ситуация.

**Распространенные типичные недочёты в проектах студентов:**

<div class="alert alert-danger" ❌>
Неработоспособный код. Проверь проект на работоспособность перед отправкой на проверку. Это важный навык, который тебе пригодится как на собеседовании, при выполнении тестового задания, так и перед совещанием у руководства компании.</div>

<div class="alert alert-danger" ❌>
Проект не структурирован. В проекте должна быть структура, т.е. названия разделов и подразделов должны быть записаны в ячейках Markdown.</div>

<div class="alert alert-warning" 👉 >
Отсутствие описания, целей проекта, его задач, выводов. В описании должны быть озвучены поставленные задачи, содержание датасета, промежуточные и финальные выводы. Как правило, заказчик читает только финальный вывод, поэтому важно добавить информацию о проведенном исследовании, применить форматирование текста, в общем, сделать красиво, чтобы было понятно, что проведено настоящее исследование и выводу можно верить.</div>  

<div class="alert alert-warning" 👉 >
"Длинный код". Код и и комментарии, которые прячутся за правым концом ячейки нужно переносить на новую строку. Твои будущие коллеги не оценят необходимость постоянного скроллинга. Согласно PEP8: https://www.python.org/dev/peps/pep-0008/ длина строки не должна превышать 79 символов.</div>

<div class="alert alert-warning" 👉 >
Твои коллеги ценят и твое и своё время, поэтому "подсвечивай" свои комментарии в ячейках Markdown. Подробнее про синтаксис Markdown, с примерами, можно почитать здесь: https://paulradzkov.com/2014/markdown_cheatsheet/?</div>


# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

## Откройте и изучите файл

In [ ]:

# Импортирование необходимых модулей и атрибутов
import pandas as pd
from joblib import dump
from sklearn.metrics import accuracy_score 
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression 
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import LinearSVC
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import Normalizer
from matplotlib import pyplot
import warnings
warnings.filterwarnings('ignore')

In [ ]:
# Загрузим датафрейм:
df = pd.read_csv('/datasets/users_behavior.csv')

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Отлично. Импортировали требуемые для анализа библиотеки и загрузили данные. Сейчас посмотрим на их состояние и если всё ОК, сможем приступать к построению моделей....
</div>

In [ ]:
# Изучим датафрейм
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [ ]:
df.describe()

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

Датафрейм содержит 3214 обьектов с 5 признаками. Пропуски в данных отстутствуют. Обьем данных достаточно мал.

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Данные загружены, отлично.

Хорошо, что применяешь метод .head()/.sample() таким образом в память не будет загружен весь датасет, объем которого может быть огромным.

Методы .info, .describe, .value_counts(), .unique() как правило часто встречаются и обязательны к применению в задачах по исследованию данных, ведь с их помощью можно получить первые инсайты. 

Итак, мы провели предварительный анализ данных, проверили данные на наличие пропусков (*из применяемых тобой методов видно что пропусков нет*) и затем обработали их (аргументированно заполнили / удалили / оставили как есть). Теперь можно приступать собственно к построению моделей и ответам на вопросы проекта.
</div>

<div class="alert alert-warning">
<b>Комментарий 👉</b>

Несколько моментов "на будущее" (сейчас забегаю чуть вперед, но будущем это будет важно) которые я хотел озвучить перед переходом к собственно исследованию:
    
- имеет смысл критично посмотреть на данные с точки зрения наличия пропусков в признаках (вопрос: все ли ml-модели "любят" пропуски?). Из результата info / describe / .duplicated() видно что пропусков нет, но я бы ещё явно это сказал. СДЕЛАНО
    
- имеет смысл явно обозначить какую задачу мы будем решать: с учителем / без учителя, классификации / регрессии ...
    
- в моделях бинарной классификации имеет смысл смотреть на баланс классов. Два взгляда (с разных сторон) на эту проблему тут: https://alexanderdyakonov.wordpress.com/2021/05/27/imbalance/ и https://habr.com/ru/post/349078/
    
- имеет смысл смотреть на корреляцию признаков. При высокой их степени коррелированности возникает проблема  мультиколлиниарность признаков. Проблема "мультиколлинеарности" это и про переобучение модели и про её интерпретируемость и про сложность модели (содержание в модели избыточного количества переменных). В чем негативный эффект сложной модели? Вот в чем: усложняется интерпретация параметров, оценки параметров ненадежны – получаются большие стандартные ошибки, которые меняются с изменением объема наблюдений, что делает модель регрессии непригодной для прогнозирования.
    
А теперь давай про интерпретируемость. Согласись - это важно, чтобы результат, который выдает модель можно было бы понять и объяснить с точки зрения предметной области. В случае включения в модель коллинеарных признаков может случиться так, что коэффициенты (да, проблема присуща моделям линейным ... регрессия среди них!) могут менять знаки и попытка объяснения предметной области с помощью этих коэффициентов будет противоречить здравому смыслу.
    
Вот классная статья. Простым языком, на примере показана проблема. А также, что важно - разобраны варианты как её (проблему) победить. Статья: https://habr.com/ru/company/akbarsdigital/blog/592493/ 
</div>

## Разбейте данные на выборки

In [ ]:
# Составим таблицы признаков и целового признака.
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra'] 

In [ ]:
# отделим 60% данных для обучающей выборки c сохранением целевого признака
features_train, features_valid_test, target_train, target_valid_test = train_test_split(
    features, target, test_size=0.4, random_state=12345 , stratify=target)


In [ ]:
# отделим 50% данных для тестовой выборки c сохранением целевого признака
features_test, features_valid, target_test, target_valid = train_test_split(
   features_valid_test, target_valid_test, test_size=0.5, random_state=12345, stratify=target_valid_test)


<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

1. Здесь хорошо, происходит фиксация random_state. Воспроизводимость результатов разбиения выборки на обучающую (тренировочную) / тестовую / валидационную мы обеспечили, значит при всех последующих запусках нашего кода подвыборки будут идентичными.
    
2. Доли размеров train/test/valid 3:1:1 - хорошо.
</div>

<div class="alert alert-warning">
<b>Комментарий 👉</b>

Я бы рекомендовал после разбиения данных на выборки посмотреть на их размеры и размерности. Метод shape для этого - идеальный помощник. "Цифры" по выборкам покажут верно ли мы произвели "разделение" данных.

p.s. так (**псевдокод**): X_train.shape[0] – «покажет» количество строк в тренировочной выборке, а X_train.shape[1] - количество столбцов в ней же. Ну а X_train.shape – выведет размерность train'а в виде кортежа с 2мя значениями (первое число – количество строк, второе – столбцов).</div>

Вывод:
Данные были разбиты по принципу 60/20/20.

Обучение пройдет на данных features_train
Валидация модели на данных features_valid
Лучшая модель по валидации будет применена на данных features_test

<div class="alert alert-warning">
<b>Комментарий 👉</b>

**Обучение пройдет на данных train_tariff_data Валидация модели на данных valid_tariff_data Лучшая модель по валидации будет применена на данных test_tariff_data** - мелочь, но названия не совпадают.
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Поправил</font>
</div>

<div class="alert alert-warning">
<b>Комментарий 👉</b>

Исключительно как вариант, а также воспользоваться советом со Стэковерфлоу: https://stackoverflow.com/questions/38250710/how-to-split-data-into-3-sets-train-validation-and-test/38251213#38251213

	train, valid, test = np.split(df.sample(frac=1, random_state=821), [int(.6*len(df)), int(.8*len(df))])

Также, разбивать данные можно и автоматизировано, используя библиотеку fast_ml (см. скрин). К этому скрину один комментарий – не факт, что эта библиотека fast_ml развернута здесь, на Хабе, НО дома, своей локальной машине, с установленной средой (например Анакондой: https://www.anaconda.com/products/distribution) ты вполне можешь этим вариантом пользоваться (после установки **!pip install fast_ml**):
</div>

![image.png](attachment:image.png)

## Исследуйте модели

Для решения проблем классификации и прогнозирования подходит целый диапазон алгоритмов.
Используем некоторый набор из них.
Для удобства вывода информации рядом с каждым алгоритмом будет проставлена аббревиатура. Необходимо отметить, что настройки каждого алгоритма принимаются по умолчанию (default), за исключением некоторых моментов, дабы предоставить равные условия.

Линейные алгоритмы:

— Логистическая регрессия / Logistic Regression (‘LR’)
— Линейный дискриминантный анализ / Linear Discriminant Analysis (‘LDA’)

Нелинейные алгоритмы:

— Метод k-ближайших соседей (классификация) / K-Neighbors Classifier (‘KNN’)
— Деревья принятия решений / Decision Tree Classifier (‘CART’)
— Наивный классификатор Байеса / Naive Bayes Classifier (‘NB’)
— Линейный метод опорных векторов (классификация) / Linear Support Vector Classification (‘LSVC’)
— Метод опорных векторов (классификация) / C-Support Vector Classification (‘SVC’)

Алгоритм искусственной нейронной сети:

— Многослойный персептрон / Multilayer Perceptrons (‘MLP’)

Ансамблевые алгоритмы:

— Bagging (классификация) / Bagging Classifier (‘BG’) (Bagging = Bootstrap aggregating)
— Случайный лес (классификация) / Random Forest Classifier (‘RF’)
— Экстра-деревья (классификация) / Extra Trees Classifier (‘ET’)
— AdaBoost (классификация) / AdaBoost Classifier (‘AB’) (AdaBoost = Adaptive Boosting)
— Градиентный boosting (классификация) / Gradient Boosting Classifier (‘GB’)

Таким образом, список ‘models’ содержит следующие модели:

<div class="alert alert-warning">
<b>Комментарий 👉</b>

Отличный список, но ты правда надеешся "выжать" что -то сильно больше чем .... чем ... думаю accuracy_score выше 80-82% ты вряд-ли "выбьешь" на наших скудных учебных данных?
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Нашел на Хабре. Решил как раз для тех, кто ...  (без опыта) ))) Убирать лишние алгоритмы не стал ради интереса. </font>
</div>

In [ ]:
# Переприсвоим данные под наши цели 
X_train = features_train
X_test = features_valid
Y_train = target_train
Y_test = target_valid



In [ ]:
# Настройка параметров оценивания алгоритма
num_folds = 10
n_estimators = 100
scoring = 'accuracy'

In [ ]:
models = []
models.append(('LR', LogisticRegression()))
models.append(('LDA', LinearDiscriminantAnalysis()))
models.append(('KNN', KNeighborsClassifier()))
models.append(('CART', DecisionTreeClassifier()))
models.append(('NB', GaussianNB()))
models.append(('LSVC', LinearSVC()))
models.append(('SVC', SVC()))
models.append(('MLP', MLPClassifier()))
models.append(('BG', BaggingClassifier(n_estimators=n_estimators)))
models.append(('RF', RandomForestClassifier(n_estimators=n_estimators)))
models.append(('ET', ExtraTreesClassifier(n_estimators=n_estimators)))
models.append(('AB', AdaBoostClassifier(n_estimators=n_estimators, algorithm='SAMME')))
models.append(('GB', GradientBoostingClassifier(n_estimators=n_estimators)))

Оцениваним эффективности каждого алгоритма выполним с помощью кросс-валидации. В итоге выводится сообщение (msg – сокр. от message), содержащее следующие сведения: имя модели в виде аббревиатуры, средняя оценка 10-кратной перекрёстной проверки на тренировочных данных (метрика ‘accuracy’), в скобках представлено среднее квадратичное отклонение (standard deviation), а также значение метрики ‘accuracy’ на тестовых данных.

In [ ]:


# Оценивание эффективности выполнения каждого алгоритма
scores = []
names = []
results = []
predictions = []
msg_row = []
for name, model in models:
    kfold = KFold(n_splits=num_folds) #random_state=seed
    cv_results = cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
    names.append(name)
    results.append(cv_results)
    m_fit = model.fit(X_train, Y_train)
    m_predict = model.predict(X_test)
    predictions.append(m_predict)
    m_score = model.score(X_test, Y_test)
    scores.append(m_score)
    msg = "%s: train = %.3f (%.3f) / test = %.3f" % (name, cv_results.mean(), cv_results.std(), m_score)
    msg_row.append(msg)
    acc = accuracy_score(Y_test, m_predict)
    print('Значение метрики ACC на валидационной выборке', model, acc)
    print(msg)

Значение метрики ACC на валидационной выборке LogisticRegression() 0.7465007776049767
LR: train = 0.734 (0.054) / test = 0.747
Значение метрики ACC на валидационной выборке LinearDiscriminantAnalysis() 0.7480559875583204
LDA: train = 0.749 (0.042) / test = 0.748
Значение метрики ACC на валидационной выборке KNeighborsClassifier() 0.7573872472783826
KNN: train = 0.740 (0.034) / test = 0.757
Значение метрики ACC на валидационной выборке DecisionTreeClassifier() 0.7216174183514774
CART: train = 0.706 (0.022) / test = 0.722
Значение метрики ACC на валидационной выборке GaussianNB() 0.7900466562986003
NB: train = 0.783 (0.032) / test = 0.790
Значение метрики ACC на валидационной выборке LinearSVC() 0.5147744945567652
LSVC: train = 0.562 (0.145) / test = 0.515
Значение метрики ACC на валидационной выборке SVC() 0.7371695178849145
SVC: train = 0.744 (0.046) / test = 0.737
Значение метрики ACC на валидационной выборке MLPClassifier() 0.713841368584759
MLP: train = 0.646 (0.138) / test = 0.714


<div class="alert alert-warning">
<b>Комментарий 👉</b>

Я правиьно понимаю, что X_test - это название валидационной выборки? А features_valid (ниже) - это название тестовой? Если да, то зачем такая путаница? ))))
</div>

<div class="alert alert-danger">
<b>Необходимо исправить ❌</b>

Что вот здесь происходит и для чего? Значение m_score - о чем нам говорит?
    
![image.png](attachment:image.png)
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Поправил</font>
</div>

<div class="alert alert-success">
<b>КОММЕНТАРИЙ V2</b>

Я всё-равно не могу понять, что ты хочешь сказать этим кодом:
    
![image.png](attachment:image.png)    
</div>

In [1]:
# Диаграмма размаха («ящик с усами»)
fig = pyplot.figure( figsize=(14, 10))
fig.suptitle('Сравнение результатов выполнения алгоритмов')
ax = fig.add_subplot(111)
red_square = dict(markerfacecolor='r', marker='s')
pyplot.boxplot(results, flierprops=red_square,)
ax.set_xticklabels(names, rotation=45)
pyplot.show()

NameError: ignored

   Лучший результат экпресс тестирования: Значение метрики ACC на валидационной выборке GradientBoostingClassifier() 0.8133748055987559
GB: train = 0.797 (0.031) / test = 0.850
  


Пропробуем добиться лучшего результата, используя алгоритимы, изученные нами в теоретической части:

исследуем логистическую регрессиию.
Изменим гипер параметр max_iter в цикле.

In [ ]:
best_max_iter = 0
best_accuracy = 0

for max_iter in range(1,51):
    model = LogisticRegression(random_state=12345, max_iter=max_iter)
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, valid_predictions)
    print('Количество итерраций', max_iter,'Точность', accuracy)
    if accuracy > best_accuracy:
        best_max_iter = max_iter
        best_accuracy = accuracy

print('Лучшая количество итерраций', best_max_iter,'Лучшая точность', best_accuracy)    

Количество итерраций 1 Точность 0.6936236391912908
Количество итерраций 2 Точность 0.6936236391912908
Количество итерраций 3 Точность 0.6936236391912908
Количество итерраций 4 Точность 0.6936236391912908
Количество итерраций 5 Точность 0.6936236391912908
Количество итерраций 6 Точность 0.6936236391912908
Количество итерраций 7 Точность 0.6936236391912908
Количество итерраций 8 Точность 0.6936236391912908
Количество итерраций 9 Точность 0.6936236391912908
Количество итерраций 10 Точность 0.6936236391912908
Количество итерраций 11 Точность 0.6936236391912908
Количество итерраций 12 Точность 0.6936236391912908
Количество итерраций 13 Точность 0.6936236391912908
Количество итерраций 14 Точность 0.6936236391912908
Количество итерраций 15 Точность 0.6936236391912908
Количество итерраций 16 Точность 0.6982892690513219
Количество итерраций 17 Точность 0.6998444790046656
Количество итерраций 18 Точность 0.7122861586314152
Количество итерраций 19 Точность 0.7076205287713841
Количество итерраций 

Лучшая количество итерраций 31 Лучшая точность 0.715. Зафиксируем:

In [ ]:
max_iter=1
model = LogisticRegression(random_state=12345, max_iter=31)
model.fit(features_train, target_train)
valid_predictions = model.predict(features_valid)
accuracy = accuracy_score(target_valid, valid_predictions)
print('Количество итерраций', max_iter,'Точность',accuracy)

Количество итерраций 1 Точность 0.7153965785381027


Изменим другой гиперпараметр - solver.

In [ ]:
best_solver = ''
best_accuracy = 0

for solve in {'newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'}:
    model = LogisticRegression(random_state=12345, solver=solve)
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, valid_predictions)
    print('Solver', solve,'Точность', accuracy)
    if accuracy > best_accuracy:
        best_solver= solve
        best_accuracy = accuracy

print('Лучший solver', best_solver,'Лучшая точность', best_accuracy) 

Solver lbfgs Точность 0.7465007776049767
Solver sag Точность 0.6936236391912908
Solver newton-cg Точность 0.7465007776049767
Solver liblinear Точность 0.7107309486780715
Solver saga Точность 0.6936236391912908
Лучший solver lbfgs Лучшая точность 0.7465007776049767


Лучшим solver является newton-cg. 

Изменим в цикле гиперпараметр С:

In [ ]:
best_c = 0
best_accuracy = 0

for float_c in [x/10 for x in range(1,21)]:
    model = LogisticRegression(random_state=12345, C=float_c)
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, valid_predictions)
    print('C=', float_c,'Точность',accuracy)
    if accuracy > best_accuracy:
        best_c = float_c
        best_accuracy = accuracy

print('Лучшая количество итерраций', best_c,'Лучшая точность', best_accuracy)    

C= 0.1 Точность 0.7465007776049767
C= 0.2 Точность 0.7465007776049767
C= 0.3 Точность 0.7465007776049767
C= 0.4 Точность 0.7465007776049767
C= 0.5 Точность 0.7465007776049767
C= 0.6 Точность 0.7465007776049767
C= 0.7 Точность 0.7465007776049767
C= 0.8 Точность 0.7465007776049767
C= 0.9 Точность 0.7465007776049767
C= 1.0 Точность 0.7465007776049767
C= 1.1 Точность 0.7465007776049767
C= 1.2 Точность 0.7465007776049767
C= 1.3 Точность 0.7465007776049767
C= 1.4 Точность 0.7465007776049767
C= 1.5 Точность 0.7465007776049767
C= 1.6 Точность 0.7465007776049767
C= 1.7 Точность 0.7465007776049767
C= 1.8 Точность 0.7465007776049767
C= 1.9 Точность 0.7465007776049767
C= 2.0 Точность 0.7465007776049767
Лучшая количество итерраций 0.1 Лучшая точность 0.7465007776049767


Значение по умолчанию C=0.1 показывает лучший результат.

Исследуем алгоритм классификации деревом решений:


In [ ]:
best_depth = 0
best_result = 0
for depth in range(1, 50):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth) # обучите модель с заданной глубиной дерева
    model.fit(features_train, target_train) # обучите модель
    predictions = model.predict(features_valid) # получите предсказания модел
    result = accuracy_score(target_valid, predictions) # посчитайте качество модели
    print("Глубина дерева, асс", depth, result)
    if result > best_result:
        best_depth = depth
        best_result = result    
print("Accuracy лучшей модели, глубина дерева:", best_result, best_depth)

Глубина дерева, асс 1 0.7293934681181959
Глубина дерева, асс 2 0.7682737169517885
Глубина дерева, асс 3 0.7993779160186625
Глубина дерева, асс 4 0.7744945567651633
Глубина дерева, асс 5 0.8118195956454122
Глубина дерева, асс 6 0.7993779160186625
Глубина дерева, асс 7 0.7978227060653188
Глубина дерева, асс 8 0.7838258164852255
Глубина дерева, асс 9 0.7947122861586314
Глубина дерева, асс 10 0.80248833592535
Глубина дерева, асс 11 0.7807153965785381
Глубина дерева, асс 12 0.7682737169517885
Глубина дерева, асс 13 0.7589424572317263
Глубина дерева, асс 14 0.7402799377916018
Глубина дерева, асс 15 0.7402799377916018
Глубина дерева, асс 16 0.7278382581648523
Глубина дерева, асс 17 0.7371695178849145
Глубина дерева, асс 18 0.7309486780715396
Глубина дерева, асс 19 0.7200622083981337
Глубина дерева, асс 20 0.7200622083981337
Глубина дерева, асс 21 0.7247278382581649
Глубина дерева, асс 22 0.7247278382581649
Глубина дерева, асс 23 0.7247278382581649
Глубина дерева, асс 24 0.7247278382581649
Глу

In [ ]:

model = DecisionTreeClassifier(random_state=12345, max_depth=5) # обучите модель с заданной глубиной дерева
model.fit(features_train, target_train) # обучите модель
predictions = model.predict(features_valid) # получите предсказания модел
result = accuracy_score(target_valid, predictions) # посчитайте качество модели
print("Глубина дерева - 5, асс", result)
    

Глубина дерева - 5, асс 0.8118195956454122


Изменм гиперпараметр criterion с gini на entropy.

In [ ]:

model = DecisionTreeClassifier(random_state=12345, max_depth=5, criterion="entropy") # обучите модель с заданной глубиной дерева
model.fit(features_train, target_train) # обучите модель
predictions = model.predict(features_valid) # получите предсказания модел
result = accuracy_score(target_valid, predictions) # посчитайте качество модели
print("Глубина дерева - 5, асс", result)
    

Глубина дерева - 5, асс 0.8087091757387247


Результат лучше стал.

Изменим splitter с best на random.

In [ ]:

model = DecisionTreeClassifier(random_state=12345, max_depth=5, splitter='random') # обучите модель с заданной глубиной дерева
model.fit(features_train, target_train) # обучите модель
predictions = model.predict(features_valid) # получите предсказания модел
result = accuracy_score(target_valid, predictions) # посчитайте качество модели
print("Глубина дерева - 5, асс", result)
    

Глубина дерева - 5, асс 0.7807153965785381


Результат стал лучше.

Рассмотрим применение модели случайный лес

In [ ]:
best_estim = 0
best_accuracy = 0

for estim in range(1,101):
    model = RandomForestClassifier(random_state=12345, n_estimators=estim)
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, valid_predictions)
    print('Количество деревьев', estim,'Точность',accuracy)
    if accuracy > best_accuracy:
        best_estim = estim
        best_accuracy = accuracy
print('Лучшее количество деревьев', best_estim,'Лучшая точность', best_accuracy)  



Количество деревьев 1 Точность 0.7402799377916018
Количество деревьев 2 Точность 0.7558320373250389
Количество деревьев 3 Точность 0.7480559875583204
Количество деревьев 4 Точность 0.7744945567651633
Количество деревьев 5 Точность 0.7713841368584758
Количество деревьев 6 Точность 0.7838258164852255
Количество деревьев 7 Точность 0.7807153965785381
Количество деревьев 8 Точность 0.7869362363919129
Количество деревьев 9 Точность 0.7947122861586314
Количество деревьев 10 Точность 0.7931570762052877
Количество деревьев 11 Точность 0.7869362363919129
Количество деревьев 12 Точность 0.7978227060653188
Количество деревьев 13 Точность 0.7900466562986003
Количество деревьев 14 Точность 0.7931570762052877
Количество деревьев 15 Точность 0.7916018662519441
Количество деревьев 16 Точность 0.7884914463452566
Количество деревьев 17 Точность 0.7916018662519441
Количество деревьев 18 Точность 0.7916018662519441
Количество деревьев 19 Точность 0.7947122861586314
Количество деревьев 20 Точность 0.799377

<div class="alert alert-danger">
<b>Необходимо исправить ❌</b>

Что здесь происходит?
    
![image.png](attachment:image.png)    
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Просмотрел по всей работе. Везде поправил.</font>
</div>

И так:

In [ ]:
    model = RandomForestClassifier(random_state=12345, n_estimators=75)
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, valid_predictions)
    print('Точность', accuracy)

Точность 0.8164852255054432


<div class="alert alert-danger">
<b>Необходимо исправить ❌</b>

Здесь так же ошибка (ОПРЕДЕЛИСЬ какая выборка у тебя тестовая, какая валидационная - сейчас ты очень вольно переключаешся между ними).
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Просмотрел по всей работе. Везде поправил.</font>
</div>

Настроим гиперпараметр criterion с gini на entropy.


In [ ]:
    model = RandomForestClassifier(random_state=12345, n_estimators=75, criterion='entropy')
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, valid_predictions)
    print('Точность', accuracy)

Точность 0.7962674961119751


<div class="alert alert-danger">
<b>Необходимо исправить ❌</b>

Та же ошибка
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Просмотрел по всей работе. Везде поправил.</font>
</div>

Точность ухудшилась.

Изменим bootstrap с True на Flase.

In [ ]:
    model = RandomForestClassifier(random_state=12345, n_estimators=75, bootstrap=False)
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, valid_predictions)
    print('Точность', accuracy)

Точность 0.7822706065318819


<div class="alert alert-danger">
<b>Необходимо исправить ❌</b>

Та же ошибка
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Просмотрел по всей работе. Везде поправил.</font>
</div>

Точность ухудшилась.

Изменим warm_start с False на True.

In [ ]:
    model = RandomForestClassifier(random_state=12345, n_estimators=24, warm_start=True)
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, valid_predictions)
    print('Точность', accuracy)

Точность 0.7978227060653188


<div class="alert alert-danger">
<b>Необходимо исправить ❌</b>

Та же ошибка
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Просмотрел по всей работе. Везде поправил.</font>
</div>

Точность не изменилась.

Выводы
Мы провели экспресс тест алгоритмов для представленных данных с большим количеством алгоритмов. Лучшим оказался ExtraTreesClassifier(). Значение метрики ACC на валидационной выборке  0.8087091757387247

В дальнейшем, были исследованы 3 алгоритма с перебором возможных гиперпараметров.
Получены следующие данные:

Точность модели решающего дерева при глубине 5 равна 0.802. 
Точность модели случайного леса при количестве деревьев 4 равна 0.62. 
Точность модели логистической регресии равна 0.74. 
Лучший результат показывает решающее древо с гиперпараметром max_depth=5 - 0.802.

<div class="alert alert-danger">
<b>Необходимо исправить ❌</b>

Итак, 
    
1. мы выше проверяли множество, около десятка моделей. В выводе описаны 3. Опиши все что проверял и сделай вывод, покажи, какая является лучшей.
    
2. твим данным доверять нельзя: все они вычислены на разных выборках, а местами ещё и даже на каких-то перекрестных (фичи от тестовой части, таргеты от валидационной). Этот раздел надо глобально переделать. Совет: не удивляй меня количеством моделей, возьми 3: логрегрессию, лес, дерево и аккуратно для них сделай вычисления.
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Просмотрел по всей работе. Везде поправил Вывод для 3 части дополнил.</font>
</div>

<div class="alert alert-success">
<b>КОММЕНТАРИЙ V2</b>

Здесь отлично: определили лучшую модель. 

После того, как лучшая модель найдена – тестируем её (можно - только её) на тестовых данных. Все прочие модели нас уже больше не интересуют. 
</div>    

## Проверьте модель на тестовой выборке

In [ ]:

model = LogisticRegression(random_state=12345, max_iter=79)
model.fit(features_train, target_train)
valid_predictions = model.predict(features_test)
accuracy = accuracy_score(target_test, valid_predictions)
print('Точность модели логистической регресиии на тестовой выборке', accuracy)

Точность модели логистической регресиии на тестовой выборке 0.744945567651633


<div class="alert alert-warning">
<b>Комментарий 👉</b>

Я правиьно понимаю, что features_valid - это название тестовой выборки? А X_test (выше) - это название валидационной? Если да, то зачем такая путаница? ))))
</div>

In [ ]:

model = DecisionTreeClassifier(random_state=12345, max_depth=8) # обучите модель с заданной глубиной дерева
model.fit(features_train, target_train) # обучите модель
predictions = model.predict(features_test) # получите предсказания модел
result = accuracy_score(target_test, predictions) # посчитайте качество модели
print("Точность модели решающего леса на тестовой выборке", result)

Точность модели решающего леса на тестовой выборке 0.80248833592535


In [ ]:
    model = RandomForestClassifier(random_state=12345, n_estimators=24)
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_test)
    accuracy = accuracy_score(target_test, valid_predictions)
    print('Точность модели случайного леса на тестовой выборке', accuracy)

Точность модели случайного леса на тестовой выборке 0.7978227060653188


Вывод п
Для решения поставленной задачи можно выбрать модель случаного леса, точность 0.79. Учитывая разницу между текущей точностью и точностью на валидационной выборке, можно сделать вывод слабой переобученности модели.

Для 100% гарантий и при учете того, что выборка не будет большой, рекомендую использовать модель случайного леса. Показатель выше требуемого на 3 пункта.

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Здесь отлично: проверили "качество" нашей лучшей модели на тестовых данных.
</div>

## (бонус) Проверьте модели на адекватность

In [ ]:
d_target_test = target_test.copy()
d_target_test = d_target_test.transform(lambda x: 0)
print("Точность модели, заполненной наиболее весомой группой '0'", accuracy_score(target_test, d_target_test))

Точность модели, заполненной наиболее весомой группой '0' 0.6936236391912908


Сделаем вывод, что наша модель имеет большую вероятность правильного прогноза, чем условно заполненная наиболее весомым значением.

<div class="alert alert-warning">
<b>Комментарий 👉</b>

Для проверки модели на адекватность можно использовать DummyClassifier (из библиотеки sckit-learn). Это один вариант. 
    
Второй вариант - сравнить самый часто встречающийся класс в наших данных (это is_ultra == 0). Таких значений 2.229 в нашем датафрейме. Всего же значений в датафрейме 3.214. Значит самый часто встречающийся класс "занимает" 69% (2229 / 3214 == 0.693528313627878). Вот мы и получили контрольные данные для сравнительной оценки, построенной нами "лучшей" модели. Наша лучшая модель должна "побить" этот скор.
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75


# Результат ревью:

Я отмечаю адекватный уровень твоей программистской подготовки. Ты используешь продвинутые концепции языка Python, библиотеки pandas и одной из основных библиотек машинного обучения sckit-learn. Это показывает твой уровень владения инструментами анализа данных и machine learning.

С комментированием никаких проблем нет. Твоих объяснений достаточно для понимания твоими коллегами хода мыслей. Ссылки из брифа проекта в нужных местах также помогают следовать за развитием событий и дают полную информацию касаемо того, что будет происходить в том или ином разделе проекта.

Мне было просто проверять твой проект, потому что он чётко структурирован, с использованием оглавления. 

Зелёным цветом я отметил и прокомментировал удачные и элегантные решения, на которые ты можешь опираться в будущих проектах.

Жёлтым цветом я выделил то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас.

Однако есть принципиальные моменты, которые не позволяют мне принять твой проект в его нынешнем состоянии:

<div class="alert alert-danger">
<b>Необходимо исправить ❌</b> 

- Есть множественный ошибки в расчете accuracy_score и предоставлении результатов (конкретных цифр) в разделе 3.

- Нет финального, завершающего вывод проекта

- Вывод в разделе "3.  Исследуйте модели" урезан.</div>

Эти моментв требуют твоего внимания. Комментарий помечены красным цветом. 

Подводя итог: после доработки обязательно проверь комментарии и выводы. Их нужно будет скорректировать. Жду доработанную версию проекта на повторное ревью.


<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'> Спасибо за советы и ссылки!!! Да, произошла какая то путаница с выбоками (((. 1. Все проверил, добавил. 2. Добавил 3.Добавил .</font>
</div>

<div class="alert alert-success">
<b>КОММЕНТАРИЙ V2</b>
    
Всё отлично, результат достигнут.
    
Один совет (также, на будущее): как попытаться улучшить полученный результат, с минимум усилий? Ответ: мы использовали train для обучения модели, а valid - для поиска лучших значений гиперпараметров. Лучшие параметры нашли. Так почему бы теперь наши модели с выбранными гиперпараметрами не обучить на *общей* (тренировочной + валидационной) выборке (pd.concat() можно использовать для объединения). Чем больше данных, тем *лучше* модели смогут обучиться (надо проверять!). И вот теперь эту дообученную модель мы уже финально проверим на тестовой выборке (test).

Но следует учесть вот какой момент: нужно быть аккуратным с подобным «улучшением», если мы кодируем или масштабируем наши выборки. Например, в следующем проекте мы обучаемся на train’е, а затем делаем transform на валидации и тесте. Если после этого объединить трейн и валид, то это будет не совсем верно.

p.s. а вообще, достижение высокого скора не такая простая задача. Много моментов может влиять (подбор гиперпараметров - всего лишь один из них). В этой конкретно задаче мне кажется, что данных маловато, а также, возможно и фичей (столбцов в наших данных). Из-за этого модели не могут точно "настроиться на данные".
</div>